### PR TITLE
add coordinates_space.py

### PR DESCRIPTION
**Pull request by WuShichao**
@ahnitz This PR is for the coordinate transformation between space-borne GW detectors and ground-based ones.

Need to finish more things:
- [x] documentation
- [x] test with .ini in PE
- [x] use some more accurate functions from Astropy
- [x] polarization test
- [x] implement LDC convention correction
- [x] SMBHB PE/likelihood test

In order to make the code easier to review, I separated parts of the code into the following PRs:

1. apply coordinate transform for `fid_params` in the `relbin` model https://github.com/gwastro/pycbc/pull/4396
2. add standard names into `parameters.py` https://github.com/gwastro/pycbc/pull/4397
3. multiband likelihood  https://github.com/gwastro/pycbc/pull/4505

### CODE REVIEW

**Comment by ahnitz** [comment]
@WuShichao Don't create a new module for this. I think it would be better to put into the existing ones.

**Comment by WuShichao** [comment]
@ahnitz I have considered this before. I think creating a standalone module for the space detector related is not a bad idea, just like what we did for PSD.

File: pycbc/coordinates_space.py
Comment (WuShichao) [reply pycbc/coordinates_space.py:31]: Actually, this function is fast, but I will check if this can be done with Astropy.

Comment (WuShichao) [reply pycbc/coordinates_space.py]: The latest doc for this module https://www.atlas.aei.uni-hannover.de/work/shichao.wu/doc/pycbc.html#module-pycbc.coordinates_space

**Comment by ahnitz** [comment]
And the orbital assumptions part should still be made clear. I think the earth orbit is fine, but the LISA part will need to be changeable.

**Comment by WuShichao** [comment]
@ahnitz OK, I will change the names of functions or variables. For LISA orbit, it depends on two functions: (1) the rotation matrix one specifies the self-rotation and LISA plane orientation; (2) the mass center's motion one specifies LISA barycenter's motion. If we want to make this more flexible, then I think we should read in this information from a file (e.g. a .hdf), this will depend on how we store these information in this file, I prefer leave this to a future PR once we implement our detector response functions.

File: pycbc/coordinates/__init__.py
Comment (ahnitz) [reply pycbc/coordinates/__init__.py:21]: Or give it a useful name

**Comment by ahnitz** [comment]
@WuShichao Please separate the change to coordinates from all the others. I think those should be reviewed separately.

**Comment by WuShichao** [comment]
@ahnitz I can separate, then other parts depending on it will be merged after than coordinates PR (otherwise their build tests will fail).

File: pycbc/coordinates/base.py
Comment (ahnitz) [inline pycbc/coordinates/base.py:143]: Don't remove this. It should explicitly say what should be imported by * which is what __all__ does. Your module should also use this feature.

File: pycbc/transforms.py
Comment (ahnitz) [inline pycbc/transforms.py:1576]: A lot of this code looks very repetitive and is doing the same type of thing at each line. Can't you compactify this by iterative over a list of the names rather than duplicating for each?

File: pycbc/transforms.py
Comment (ahnitz) [inline pycbc/transforms.py:1810]: Can't you simplify this by using a dictionary of the defaults? And then in your init you don't need separate if statements, you can just iterate over a list? It's the same comment for each transform so I won't repeat it for each.

File: pycbc/coordinates/space.py
Comment (ahnitz) [reply pycbc/coordinates/space.py:32]: As mentioned, we should have  unittests for these functions that verify where possibly that they give answers we've validated to be correct and that they are revsesible. Make a new file in the test directory modeled after the test_conversions.py one. It will be automatically run by the test suite. You should first verify it works on your local machine though.

**Review by ahnitz — COMMENTED** [review]
@WuShichao I think this is getting in good shape. See my few comments on documentation and unit tests as you are adding a large chunk of code here. Once those are addressed, I think we can merge.

File: pycbc/coordinates/base.py
Comment (WuShichao) [reply pycbc/coordinates/base.py:143]: @ahnitz Will this be duplicated? Because `__init__.py` has `__all__` already.

File: pycbc/coordinates/space.py
Comment (WuShichao) [reply pycbc/coordinates/space.py:34]: @ahnitz I add the comment on https://github.com/gwastro/pycbc/pull/4289/files#diff-7a1da599956f99d94f2c70707533b585de4d61399182d8b58954fe73ed1e15dcR27

File: pycbc/transforms.py
Comment (WuShichao) [inline pycbc/transforms.py:1603]: @ahnitz I have a question here, you can see the default names of `lisa` and `ssb` are the same; will this cause some potential issues if the user doesn't set custom names for them? Last time in PE slack channel, you both mentioned not hard code det-related names.

File: pycbc/coordinates/base.py
Comment (ahnitz) [inline pycbc/coordinates/base.py:17]: I'm not sure this really clarifies anything? What is meant here? What kind of coordinate transformations are in this module?

File: pycbc/transforms.py
Comment (ahnitz) [reply pycbc/transforms.py:1607]: Yes, it might. Perhaps we should require that the user set these, rather than make an assumption? If we do find a convention that allows an obvious default we can always put one in.

**Comment by ahnitz** [comment]
@WuShichao I couple of the CC comments can be fixed, but some of the ones about snakecase / duplication in this case I think cannot easily. I'd address the ones you can, and then double check your unittests. Once the unittest are in place that at least excersize the key functions, I think this can be merged.

**Review by ahnitz — APPROVED** [review]
@WuShichao Provisionally approved, see the other comments, but once those are addressed, I think we can merge this.

File: pycbc/coordinates/space.py
Comment (ahnitz) [inline pycbc/coordinates/space.py:647]: Why are you looping over num here rather than just doing vector / ndarray operations as appropriate? This would be a much slower implementation if you want to calculate for many points.

**Review by ahnitz — DISMISSED** [review]

File: pycbc/coordinates/space.py
Comment (ahnitz) [inline pycbc/coordinates/space.py:628]: Why do you have these copies? That is only  needed if some function you call modifies the vector in place. Otherwise, you don't need this. Most numpy operations  return a new array.

File: test/test_coordinates_space.py
Comment (ahnitz) [inline test/test_coordinates_space.py:315]: Can you add some inline documentation here to explain this test a bit more? It's a bit opaque to someone  not already familiar with what you are trying to do. 

Separately, where do these random numbers come from, for example, how would one update these if needed? It would be good if this is also clear in the inline docs

File: pycbc/coordinates/space.py
Comment (ahnitz) [inline pycbc/coordinates/space.py:573]: Is the 'use_astropy' option now working either way? If so, maybe this should be part of the unittest to make sure given it gave so many troubles before? Or should we settle on always using astropy?

**Review by ahnitz — COMMENTED** [review]
@WuShichao I think this PR is ready to go barring the last couple comments and getting the test suite to pass (the priority). Once the test suite passes, give me a final ping and we can merge.

File: pycbc/coordinates/space.py
Comment (WuShichao) [reply pycbc/coordinates/space.py:573]: @ahnitz In the unittest, using astropy is the default.

File: test/test_coordinates_space.py
Comment (WuShichao) [reply test/test_coordinates_space.py:326]: @ahnitz I have added a paragraph, please check.

File: pycbc/coordinates/space.py
Comment (ahnitz) [reply pycbc/coordinates/space.py:658]: @WuShichao This issue actually still needs to be addressed.

**Comment by WuShichao** [comment]
@ahnitz In order to make `unittest` work, we also need to update the waveform plugin https://github.com/gwastro/BBHX-waveform-model/pull/1

**Comment by ahnitz** [comment]
@WuShichao What is the status of this PR? It seems you need to rebase and address the CC issues still in addition to fixing any remaining tests (assuming they don't pass on a rebase).

**Comment by WuShichao** [comment]
@ahnitz There is still an issue related to the BBHx setup. Other than that, I'm waiting for Connor's tests for SMBHB and debugging the `MultibandRelativeTimeDom`, currently I'm testing it within this PR. Maybe we can separate it as a new PR.

File: pycbc/inference/models/hierarchical.py
Comment (ahnitz) [inline pycbc/inference/models/hierarchical.py:697]: Don't explicitly call things ground / space-based

File: pycbc/inference/models/hierarchical.py
Comment (ahnitz) [inline pycbc/inference/models/hierarchical.py:722]: I think this may not work correctly. You should only update the variable parameters of the primary model.

File: pycbc/inference/models/hierarchical.py
Comment (ahnitz) [inline pycbc/inference/models/hierarchical.py:773]: Set primarly model explicitly in the config file. 

[mdoel]
....
primary_model = ce 
other_models = model2 model3

File: pycbc/inference/models/hierarchical.py
Comment (ahnitz) [inline pycbc/inference/models/hierarchical.py:636]: You shouldn't need that check

File: pycbc/inference/models/hierarchical.py
Comment (ahnitz) [inline pycbc/inference/models/hierarchical.py:645]: You don't need to call this.

File: pycbc/inference/models/hierarchical.py
Comment (ahnitz) [inline pycbc/inference/models/hierarchical.py:647]: You do need to take out marginalized parameters from your variable params and from the prior. The eaisest would be to just copy from the primary model, don't recalculate.

File: pycbc/inference/models/hierarchical.py
Comment (WuShichao) [reply pycbc/inference/models/hierarchical.py:721]: There are no errors from here in my test runs, but I will ensure that.

**Comment by WuShichao** [comment]
> @WuShichao Can you confirm that if the plugin PR, the tests that are currently failing here will work? (e.g. check that there are not unrelated failures). If that is the case, we can merge the plugin PR first and then rerun the tests here and then merge this when they pass.

@ahnitz Yes, we need to review and merge the waveform plugin PR first, because you can see one of the failed tests is caused by not setting the LISA position correctly https://github.com/gwastro/pycbc/actions/runs/6511157215/job/17688598659?pr=4289#step:6:192 so in our polarization test, LISA and ET can not align correctly, I have run the same test (with plugin PR) in a fresh env on ATLAS, it works:
image

File: test/test_coordinates_space.py
Comment (WuShichao) [reply test/test_coordinates_space.py:272]: Thanks for the link; this `test_coordinates_space.py` is just testing for my transform, which overlaps LISA's X channel with ET's E1, when long wavelength approximation holds for the same signal, two V-shape detectors should have very similar detector response, so this can check if my frame transform is correct or not. It's not very matter here, just a test.

File: test/test_coordinates_space.py
Comment (WuShichao) [reply test/test_coordinates_space.py:28]: This part randomly generates parameters set from those ranges, to see if my transform functions are reversible or not because LISA's orbit is periodic, so it's OK to choose a random value. Another reason is that this range covers the GPS time when 3G detectors will be online if I remember correctly.

File: test/test_transforms.py
Comment (WuShichao) [reply test/test_transforms.py:47]: Those two are just wide enough time ranges to make sure it works for "any choosen time", also this `tc` is in the SSB frame, not the geocentric one `tc_geo`.

Comment (ConWea) [reply test/test_transforms.py:47]: Thanks for the explanation!

**Review by ConWea — APPROVED** [review]

**Pull request merged**
